main/src/main/scala/sbt/internal/server/BuildServerReporter.scala

```diff
@@ -28,110 +26,90 @@ class BuildServerReporter(
-    aggregateProblems(problem)
+    publishDiagnostic(problem)
 
     // console channel can keep using the xsbi.Problem
     super.logInfo(problem)
   }
 
-  private[sbt] def resetPrevious(analysis: CompileAnalysis): Unit = {
-    val files = analysis.readSourceInfos.getAllSourceInfos.keySet.asScala
-    files foreach { file =>
+  private def publishDiagnostic(problem: Problem): Unit = {
+    for {
+      sourceFile <- problem.position.sourceFile.toOption
+      diagnostic <- toDiagnostic(problem)
+    } {
       val params = PublishDiagnosticsParams(
-        TextDocumentIdentifier(converter.toPath(file).toUri),
+        TextDocumentIdentifier(sourceFile.toURI),
         buildTarget,
-        None,
-        diagnostics = Vector(),
-        reset = true
+        originId = None,
+        Vector(diagnostic),
+        reset = false
       )
       exchange.notifyEvent("build/publishDiagnostics", params)
     }
   }
 
-  private[sbt] def aggregateProblems(problem: Problem): Unit = {
+  private def toDiagnostic(problem: Problem): Option[Diagnostic] = {
     val pos = problem.position
-    pos.sourcePath().toOption foreach { sourcePath: String =>
-      problemsByFile.get(sourcePath) match {
-        case Some(xs: mutable.ListBuffer[Problem]) =>
-          val diagnostics = toDiagnostics(xs)
-          val absolutePath = converter.toPath(new BasicVirtualFileRef(sourcePath) {})
-          val params = PublishDiagnosticsParams(
-            TextDocumentIdentifier(absolutePath.toUri),
-            buildTarget,
-            originId = None,
-            diagnostics,
-            reset = true
-          )
-          exchange.notifyEvent("build/publishDiagnostics", params)
-        case _ =>
-      }
-    }
-  }
-
-  private[sbt] def toDiagnostics(ps: Seq[Problem]): Vector[Diagnostic] = {
     for {
-      problem <- ps.toVector
-      pos = problem.position
-      line0 <- pos.line.toOption.toVector
-      pointer0 <- pos.pointer.toOption.toVector
+      line <- pos.line.toOption.map(_.toLong - 1L)
+      pointer <- pos.pointer.toOption.map(_.toLong)
     } yield {
-      val line = line0.toLong - 1L
-      val pointer = pointer0.toLong
-      val range = (
-        pos.startLine.toOption,
-        pos.startColumn.toOption,
-        pos.endLine.toOption,
-        pos.endColumn.toOption
-      ) match {
-        case (Some(sl), Some(sc), Some(el), Some(ec)) =>
-          Range(Position(sl.toLong - 1, sc.toLong), Position(el.toLong - 1, ec.toLong))
-        case _ =>
-          Range(Position(line, pointer), Position(line, pointer + 1))
-      }
+      val range = for {
+        sl <- pos.startLine.toOption
+        sc <- pos.startColumn.toOption
+        el <- pos.endLine.toOption
+        ec <- pos.endColumn.toOption
+      } yield Range(Position(sl.toLong - 1, sc.toLong), Position(el.toLong - 1, ec.toLong))
+
       Diagnostic(
-        range,
+        range.getOrElse(Range(Position(line, pointer), Position(line, pointer + 1))),
```

Comment:
`for` desugars into flatMap, and it's going to be expensive when there are nested ones like this. I think the original pattern match was clear enough.